### ISSUE TITLE
Use Composer autoload-dev section for autoloading classes in unit tests

### ISSUE THREAD

**Issue by arogachev** [yiisoft/yii2#12943]
During working on #12840 I discovered [autoload-dev](https://getcomposer.org/doc/04-schema.md#autoload-dev) section of Composer.

> `autoload-dev` (root-only)#
>
> This section allows to define autoload rules for development purposes.
>
> Classes needed to run the test suite should not be included in the main autoload rules to avoid polluting the autoloader in production and when other people use your package as a dependency.
>
> Therefore, it is a good idea to rely on a dedicated path for your unit tests and to add it within the `autoload-dev` section.
>
> Example:
>
> ```json
> {
>     "autoload": {
>         "psr-4": { "MyLibrary\\": "src/" }
>     },
>     "autoload-dev": {
>         "psr-4": { "MyLibrary\\Tests\\": "tests/" }
>     }
> }
> ```

We can add this to `composer.json`:

```json
"autoload-dev": {
    "psr-4": {
        "yiiunit\\": "tests/"
    }
},
```

and run `composer dump-autoload`.

Then this line in `tests/bootstrap.php` will become unnecessary:

```php
Yii::setAlias('@yiiunit', __DIR__);
```

Also want to ask about this block:

```php
// require composer autoloader if available
$composerAutoload = __DIR__ . '/../vendor/autoload.php';
if (is_file($composerAutoload)) {
    require_once($composerAutoload);
} else {
    require_once(__DIR__ . '/../../../autoload.php');
}
```

- Running `composer update` is a required condition in [Prepare testing environment](https://github.com/yiisoft/yii2/blob/master/docs/internals/git-workflow.md#3-prepare-the-testing-environment-) section of [Git workflow for Yii 2 contributors](https://github.com/yiisoft/yii2/blob/master/docs/internals/git-workflow.md).
- I think Composer is used in Travis too,

So is this check really needed?

Note that to ignore `autoload-dev` section on production, we need to use `--no-dev` flag (found it not in official docs, but in this [blog article](https://hannesvdvreken.com/2015/01/28/autoload-dev/)). We can cover this in docs.

For example in [Deployer](https://github.com/deployphp/deployer) deploy tool it's already considered in [common recipe](https://github.com/deployphp/deployer/blob/master/recipe/common.php):

```php
set('composer_options', '{{composer_action}} --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader');
```

[Laravel](https://github.com/laravel/laravel/blob/master/composer.json) already uses it for tests.

So seems like this is a recommended way of handling autoloading in test environment in case of using Composer (and we use it for tests).

I can send a PR if you agree with this change.

**Comment by SilverFire** [comment]
Thank you for discovering. We are waiting for the PR

**Comment by cebe** [comment]
> Then this line in tests/bootstrap.php will become unnecessary:
> 
> Yii::setAlias('@yiiunit', __DIR__);

this is not true, the alias is not only used for autoloading but for file paths used inside unit tests.
Also it adds an additional step to get unit tests to work, which needs to be repeated whenever a class is added. using composer autoloader here is less flexible than using yii's.

> Also want to ask about this block:
> [...]
> So is this check really needed?

no, this is copy paste from old tests file structure, fixed it: e675d5a the check is still useful as most of the code should work without a composer install.

> Note that to ignore autoload-dev section on production, we need to use --no-dev flag (found it not in official docs, but in this blog article). We can cover this in docs.

This is not needed, autoload-dev will not affect installation of yii via composer. it will only affect you when you clone the yii2 repo directly.
